### ISSUE TITLE
1.13.8 (Ubuntu 16.10) Liberty: Hide and Seek bug?

### ISSUE THREAD

**Issue by GuidoBartoli** [wesnoth/wesnoth#1955]
Hi,
I'm having a strange issue with the _Hide and Seek_ scenario of the **Liberty** campaign.
The main focus is to not be intercepted by enemy patrols, and once they saw you, both heroes are practically doomed against level 3 soldiers.

During a game, I made a mistake and the main character has been killed so I had to take back some turns to try avoid this. However, even if I reload previous turns, even if I did not yet enter into enemy line sights, they are already aware of my units' presence and they simply attack them even if they are out of their range. Rolling back to previous turns, the mission could not be completed anymore, the only way is to restart it from the beginning.

I have attached my save games, so I hope someone can reproduce this problem.
[Hide_And_Seek.zip](https://github.com/wesnoth/wesnoth/files/1287036/Hide_And_Seek.zip)

**Comment by hrubymar10** [comment]
Hi @GuidoBartoli ,
I can't really remember 'how it should work'. Help me remember. Evil units should just 'stay' on their positions and didn't move at all? If so than yes, I can reproduce it in latest master:HEAD.

```
The Battle for Wesnoth version 1.13.8+dev (3e9515c1d4-Modified)
Running on Apple macOS 10.13

Game paths
==========

Data dir:        /Users/user/xcode/wesnoth/projectfiles/Xcode/build/Release/Wesnoth.app/Contents/Resources
User config dir: /Users/user/Library/Application Support/Wesnoth_1.13
User data dir:   /Users/user/Library/Application Support/Wesnoth_1.13
Saves dir:       /Users/user/Library/Application Support/Wesnoth_1.13/saves
Add-ons dir:     /Users/user/Library/Application Support/Wesnoth_1.13/data/add-ons
Cache dir:       /Users/user/Library/Application Support/Wesnoth_1.13/cache

Libraries
=========

Boost:             1.65
OpenSSL/libcrypto: 1.0.2l  (runtime 1.0.2l)
Cairo:             1.14.10 (runtime 1.14.10)
Pango:             1.40.12 (runtime 1.40.12)
SDL:               2.0.5   (runtime 2.0.5)
SDL_image:         2.0.1   (runtime 2.0.1)
SDL_mixer:         2.0.1   (runtime 2.0.1)
SDL_ttf:           2.0.14  (runtime 2.0.14)
libpng:            1.6.32  (runtime 1.6.32)

Features
========

Experimental OpenMP support:    no
PNG screenshots:                yes
Lua console completion:         yes
Legacy bidirectional rendering: no
Cocoa notifications back end:   yes
Growl notifications back end:   no
```

**Comment by ln-zookeeper** [comment]
Thanks, this is a very useful report in that it details how to reproduce it. I've come across what I believe is the same problem before elsewhere, but I didn't know that loading a previous save was required to trigger it.

It seems the root cause is that some AI customizations aren't saved properly in savefiles, causing the AI behavior to change as soon as you load a mid-scenario save, in this case causing the guards on the side that is supposed to have an idle AI to start attacking you.

**Comment by ln-zookeeper** [comment]
A few more details that might be helpful:

Firstly, reproducing is as simple as starting the Hide and Seek scenario (:cl from campaign start works fine), and then immediately loading the turn 1 autosave.

Secondly, this does seem to only affect idle_ai sides. I haven't tested it very precisely, though.

Thirdly, here's some of the resulting differences in the AI configs before and after loading, as seen through :inspect -> team 2 -> ai -> engines:

After starting the scenario, when everything is working correctly:
```
[engine]
	engine="cpp"
	name="cpp"
[/engine]
[engine]
	engine="cpp"
	name="fai"
	[formula_ai]
	[/formula_ai]
[/engine]
```

After loading a save (such as the turn 1 autosave):
```
[engine]
	engine="cpp"
	name="cpp"
[/engine]
[engine]
	code="wesnoth.require(""ai/lua/dummy_engine_lua.lua"")"
	engine="cpp"
	id=""
	name="lua"
	[data]
	[/data]
[/engine]
[engine]
	engine="cpp"
	name="fai"
	[formula_ai]
	[/formula_ai]
[/engine]
```

After loading, :inspect -> team 2 -> ai -> stages also contains an ai_default_rca main_loop (whereas it was and should remain empty), and there might be other differences in aspects and component structure.

**Comment by GuidoBartoli** [comment]
Yep, it seems a sort of permanent flag enabling attacks that remains activated even if it was off in the saved game...

**Comment by CelticMinstrel** [comment]
Just for the record, the difference in the engines has nothing to do with this - the Lua engine is always implicitly added on demand, unless it was added explicitly.

**Comment by gfgtdf** [comment]
@CelticMinstrel could you please fix this? Iirc you are the only one who changed the ai loading code after 1.12 . So you are probably not only the only  one who knows that code you are also very likeley the one who broke this one.

**Comment by CelticMinstrel** [comment]
From what Zookeeper said it sounds more likely to be an issue in the AI _saving_ code, which I don't recall changing (or at least not much). I can look into it at some point, though.

**Comment by gfgtdf** [comment]
> From what Zookeeper said it sounds more likely to be an issue in the AI saving code

zookeeper did some testing in irc wesnoth-dev (yesterday iirc?), and to me it sounded lieka bug in the loading code.

**Comment by gfgtdf** [comment]
marking this a blcker since it breaks a mainline campaign and probably other umc/mainline scenarios too.

**Comment by mattsc** [comment]
Has anything been done about this recently?  I just tried ln-zookeeper's method and could not reproduce the error with v1.13.10 (0c2d3a630e-Clean).

For the record, it is the difference in the stages that will cause this, the differences in engines, aspects etc. don't matter, but the side 2 stage remains empty after reloading in my case, so I cannot try to debug this.

**Comment by mattsc** [comment]
I figured it out and committed the fix in cfb2aa8. It was surprisingly simple. See the commit message for the cause of the problem. You can also check out the wesnoth-dev irc channel logs starting at 20171002 20:23:54 for a bit more on the testing and discussion we did to find this.

**Comment by GuidoBartoli** [comment]
Nice to hear, good work! :+1:
